### ISSUE TITLE
Better slots handling in TaClassCompositionElement >> slots

### ISSUE THREAD

**Issue by MarcusDenker** [guillep/pharo-core#3042]
From: https://pharo.fogbugz.com/f/cases/21972
based on discussion in https://github.com/pharo-project/pharo/pull/1253

dionisiydk on Apr 24  •  Member
It was here before. But now I noticed that "TraitedClass slots" is empty.
Does this logic really needed?

And what will happens if I will have class with var names equal to one from TraitedClass? And then I will add trait to it. Relying on name does not looks nice.

  @tesonep
tesonep on Apr 25  Member
Nowadays we don't have instance variables in the TraitedClass, but removing it will limit the future ability. Maybe we should rewrite the method to not do the reject if the TraitedClass slots is empty.

  @dionisiydk
dionisiydk on Apr 25  Member
But what about my scenario in case when TraitedClass has var:

And what will happens if I will have class with var names equal to one from TraitedClass? And then I will add trait to it.

Is it normal conflict as other trait conflicts? Then resolving it with alias will break this rejection logic. Does it?
Should not it be like:

innerClass slots reject: [:e | TraitedClass slots includes: e originSlot] "like originMethod"
But I don't know what is innerClass here. So maybe it is all wrong

  @dionisiydk
dionisiydk on Apr 25  Member
Could you explain why TraitedClass slots should be rejected?

  @tesonep
tesonep on Apr 25  Member
Because they are handled in different way.
Basically, including them depends if they are in the superclass or not. As this trait is included in all the traited class, if I inherit from a traited class (the slots and methods) are already in the superclass and should not be duplicated. However, if the new class inherits from a normal class the methods and slots should be added.

  @dionisiydk
dionisiydk on Apr 25  Member
Ok. I see complexity.
But do you agree that checking slots by name can lead to problems?
